component/src/ibc/transfer.rs

```diff
@@ -28,6 +35,71 @@ impl ICS20Transfer {
     pub fn new(state: State) -> Self {
         Self { state }
     }
+
+    pub async fn withdrawal_check(&self, ctx: Context, withdrawal: &ICS20Withdrawal) -> Result<()> {
+        // create packet
+        let packet: IBCPacket<Unchecked> = withdrawal.clone().into();
+
+        // send packet
+        use crate::ibc::packet::SendPacket;
+        self.state.send_packet_check(ctx.clone(), packet).await?;
+
+        Ok(())
+    }
+
+    pub async fn withdrawal_execute(&mut self, ctx: Context, withdrawal: &ICS20Withdrawal) {
+        // create packet
+        let packet: IBCPacket<Unchecked> = withdrawal.clone().into();
+
+        use crate::ibc::packet::SendPacket;
+        let checked_packet = self
+            .state
+            .send_packet_check(ctx.clone(), packet)
+            .await
+            .unwrap();
+
+        // TODO: we should probably encapsulate this in a type
+        let prefix = format!("{}/{}/", withdrawal.source_port, withdrawal.source_channel);
+        // we are the source if the denomination is not prefixed with the source port and source
+        // channel
+        let is_source = !withdrawal.denom.starts_with(&prefix);
```

Comment:
We should probably encapsulate this checking logic in a type. We could also use Penumbra's chain state to record a list of all IBC assets and reference that list, instead of doing string parsing, but that might be redundant